go/vt/discovery/endpointwatcher.go

```diff
@@ -0,0 +1,104 @@
+package discovery
+
+import (
+	"time"
+
+	log "github.com/golang/glog"
+	pbt "github.com/youtube/vitess/go/vt/proto/topodata"
+	"github.com/youtube/vitess/go/vt/topo"
+	"golang.org/x/net/context"
+)
+
+// NewEndPointWatcher returns an EndPointWatcher and starts refreshing.
+func NewEndPointWatcher(topoServer topo.Server, hc HealthCheck, cellsToWatch []string, refreshInterval time.Duration) *EndPointWatcher {
+	epw := &EndPointWatcher{
+		topoServer:      topoServer,
+		hc:              hc,
+		cellsToWatch:    cellsToWatch,
+		refreshInterval: refreshInterval,
+		endPoints:       make(map[string]*pbt.EndPoint),
+	}
+	epw.start()
+	return epw
+}
+
+// EndPointWatcher pulls endpoints of all running tablets periodically.
+type EndPointWatcher struct {
+	topoServer      topo.Server
+	hc              HealthCheck
+	cellsToWatch    []string
+	refreshInterval time.Duration
+	cancelFunc      context.CancelFunc
+	endPoints       map[string]*pbt.EndPoint
+}
+
+// start starts the pulling.
+func (epw *EndPointWatcher) start() {
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	epw.cancelFunc = cancelFunc
+	go epw.watch(ctx)
+}
+
+// watch pulls all endpoints and notifies HealthCheck by adding/removing endpoints.
+func (epw *EndPointWatcher) watch(ctx context.Context) {
+	ticker := time.NewTicker(epw.refreshInterval)
```

Comment:
I see to remember this will sleep for (interval) before kicking in for the first time. Might delay start-up too much? Break up the function to call when the timer hit, and you can call it at the beginning, then start the timer.